allennlp/models/bidaf_self_attention.py

```diff
@@ -0,0 +1,303 @@
+
+        encoded_question = self._dropout(self._phrase_layer(embedded_question, question_lstm_mask))
+        encoded_passage = self._dropout(self._phrase_layer(embedded_passage, passage_lstm_mask))
+        encoding_dim = encoded_question.size(-1)
+
+        # Shape: (batch_size, passage_length, question_length)
+        passage_question_similarity = self._matrix_attention(encoded_passage, encoded_question)
+        # Shape: (batch_size, passage_length, question_length)
+        passage_question_attention = util.last_dim_softmax(passage_question_similarity, question_mask)
+        # Shape: (batch_size, passage_length, encoding_dim)
+        passage_question_vectors = util.weighted_sum(encoded_question, passage_question_attention)
+
+        # We replace masked values with something really negative here, so they don't affect the
+        # max below.
+        masked_similarity = util.replace_masked_values(passage_question_similarity,
+                                                       question_mask.unsqueeze(1),
+                                                       -1e7)
+        # Shape: (batch_size, passage_length)
+        question_passage_similarity = masked_similarity.max(dim=-1)[0].squeeze(-1)
+        # Shape: (batch_size, passage_length)
+        question_passage_attention = util.masked_softmax(question_passage_similarity, passage_mask)
+        # Shape: (batch_size, encoding_dim)
+        question_passage_vector = util.weighted_sum(encoded_passage, question_passage_attention)
+        # Shape: (batch_size, passage_length, encoding_dim)
+        tiled_question_passage_vector = question_passage_vector.unsqueeze(1).expand(batch_size,
+                                                                                    passage_length,
+                                                                                    encoding_dim)
+
+        # Shape: (batch_size, passage_length, encoding_dim * 4)
+        final_merged_passage = torch.cat([encoded_passage,
+                                          passage_question_vectors,
+                                          encoded_passage * passage_question_vectors,
+                                          encoded_passage * tiled_question_passage_vector],
+                                         dim=-1)
+
+        final_merged_passage = F.relu(self._merge_atten(final_merged_passage))
+
+        residual_layer = self._dropout(self._residual_encoder(self._dropout(final_merged_passage), passage_mask))
+        self_atten_matrix = self._self_atten(residual_layer, residual_layer)
+
+        mask = passage_mask.resize(batch_size, passage_length, 1) * passage_mask.resize(batch_size, 1, passage_length)
+
+        # torch.eye does not have a gpu implementation, so we are forced to use the cpu one and .cuda()
+        # Not sure if this matters for performance
+        self_mask = Variable(torch.eye(passage_length, passage_length).cuda()).resize(1, passage_length, passage_length)
```

Comment:
This will not work on machines without gpus, which is annoying for developers without gpus on their local machine.